lib/puppet/state_machine/machine_builder.rb

```diff
@@ -0,0 +1,182 @@
+          opts[:transitions].each do |event, target|
+            trans[event] = target
+          end
+        end
+        @states[state_name] = State.new(state_name, opts[:action], opts[:event], opts[:type], trans)
+      end
+
+      # Add a transition between states.
+      #
+      # @note The source state must be defined before defining transitions for that state.
+      #
+      # @example
+      #   machine.transition(
+      #     source: :create_lockfile,
+      #     target: :run_agent,
+      #     on: :got_lock)
+      #
+      # @return [Hash<Event, Target>] The defined transitions for the named source state
+      def transition(opts)
+        source = opts[:source]
+        target = opts[:target]
+        event = opts[:on]
+        source_state = @states[opts[:source]]
+
+        if source.nil?
+          raise ArgumentError, "Unable to add transition: source state not defined"
+        end
+
+        if target.nil?
+          raise ArgumentError, "Unable to add transition: target state not defined"
+        end
+
+        if event.nil?
+          raise ArgumentError, "Unable to add transition: transition event not defined"
+        end
+
+        if source_state.nil?
+          raise ArgumentError, "Unable to add transition '#{opts[:source]}' to '#{opts[:target]}': source state '#{opts[:source]}' is not defined"
+        end
+
+        source_state.transitions[event] = target
+        source_state.transitions
+      end
+
+      # Mark a state as final.
+      #
+      # Composing state machines strips the final status from states so that the machines can be
+      # composed; states that should be final in the composed machine must be individually marked
+      # as final with this method.
+      def final(*state_names)
+        state_names.each do |state_name|
+          s = @states[state_name]
+          if s.nil?
+            raise ArgumentError, "Cannot mark state '#{state_name}' as final: state not defined"
+          else
+            s.type = :final
+          end
+        end
+      end
+
+      private
+
+      def verify_start_state!
+        if @start_state.nil?
+          raise ArgumentError, "Unable to build state machine '#{@machine_name}': start state not named"
+        elsif @states[@start_state].nil?
+          raise ArgumentError, "Unable to build state machine '#{@machine_name}': start state '#{@start_state}' not defined"
+        end
+      end
+
+      def verify_transition_targets!
+        @states.values.each do |state|
+          state.transitions.each_pair do |event, target|
+            if @states[target].nil?
+              raise ArgumentError, "Unable to build state machine '#{@machine_name}': invalid transition from from state '#{state.name}' to missing state '#{target}'"
+            end
+          end
+        end
+      end
+
+      def verify_exiting_transitions!
+        @states.values.each do |state|
+          if !(state.error? || state.final?) && state.transitions.empty?
```

Comment:
I'm guessing this is overkill but it seems like the first half of this, ie `!(state.error? || state.final?)` is really asking `is this is a terminal state?` which might be logic better encapsulated in a method on the State class, ie `#is_terminal_state?`